### ISSUE TITLE
Improve minMaxHelpers in z/codegen

### ISSUE THREAD

**Issue by dchopra001** [eclipse/omr#2969]
The maxMinHelper routine in the z code generator handles tree evaluation for `max` and `min` nodes for integer values. 
https://github.com/eclipse/omr/blob/4cc22af925fa2c0738d3adfd4b5e53093b57c3c0/compiler/z/codegen/ControlFlowEvaluator.cpp#L911-L923

There are some improvements we can make to this evaluator:

1. Write tril tests so we capture all corner cases. (One such case was fixed here 
https://github.com/eclipse/omr/pull/2967#issuecomment-421469732. For the future, we would like to have tests that verify all corner cases are handled correctly). The current approach uses signed comparisons for the first compare, and a logical compare for the second compare. To see if this approach will always produce the correct result, we can write into a table all the cases for the upper and lower halves:
```
A < B, A == B, A > B for upper and X < Y, X == Y, X > Y for lower 32-bit halves
```

Then we can write tril tests to capture all those cases in a unit test.

2. For 32-bit compilers, we can improve the evaluator to use a branchless sequence. Currently for 32-bit compilers if we have to use register pairs, we do the following:

https://github.com/eclipse/omr/blob/4cc22af925fa2c0738d3adfd4b5e53093b57c3c0/compiler/z/codegen/ControlFlowEvaluator.cpp#L943-L951

For the following example:
```
(M:N) = lmin((A:B), (X:Y))
M - high word of result register, N - low word of result register
A - high word of first value's register, B - low word of first value's register
X - high word of second value's register, Y - low word of second value's register
```

We generate the instructions below:
```
CR A,X    ; signed comparison of the two high words
LOCR M, X  ; conditionally load X into M (the result register's high word) if CC=2
LOCR N, Y   ; do the same for the low word
BRC COND_BNE ; Branch to endLabel if A,X are NOT equal if they're are equal, then we don't know max/min yet)
CLR B, Y    ; signed comparison of the two low words
LOCR M, X  ; conditionally load X into M (the result register's high word) if CC=2
LOCR N, Y   ; do the same for the low word
endLabel
```

We can improve the above sequence by removing the branch instructions. Branches around conditional loads are very expensive. Something like this can be done to avoid branches altogether:
```
(M:N) = lmin((A:B), (X:Y))

CR B,Y
LOCR M,X, CC=2
LOCR N,Y, CC=2
CR A,X
LOCR M,A, CC=1
LOCR N,B, CC=1
```

Or a similar approach using `CLR` to do the second comparison can be used as well.

3. The `LOCR` instructions used in this evaluator require a minimum ALS of z196. We need an alternative path if we are on z9 or z10. Either that, or we need a fatal assert in the evaluator on the required processor and fix any downstream code to not transform to min/max nodes if on z10 or below (the second approach keeps things simpler).

**Comment by dchopra001** [comment]
@dhong44 : for you :)

**Comment by dhong44** [comment]
Regarding point 2 It seems like the branchless sequence outlined is incorrect. When `X > A` and `Y ≤ B`, we would never load (X:Y) into (M:N). I think in this case, an extra compare and conditional load is necessary. A working branchless sequence of instructions is:

```
(M:N) = lmin((A:B), (X:Y))

LR N,B
LR M,A

CR N,Y
LOCR M,X, CC=2
LOCR N,Y, CC=2

CR M,B
LOCR M,A, CC=2
LOCR N,B, CC=2

CR M,X
LOCR M,X, CC=2
LOCR N,Y, CC=2
```

**Comment by fjeremic** [comment]
> Regarding point 2 It seems like the branchless sequence outlined is incorrect. When `X > A` and `Y ≤ B`, we would never load (X:Y) into (M:N).

Let's include this with a counterexample and write a unit test for it.

**Comment by dchopra001** [comment]
Fixed in #2980.
